### ISSUE TITLE
[Question] Does flynn use libvirt-lxc instead of sending requests to dockerd?

### ISSUE THREAD

**Issue by resouer** [flynn/flynn#1421]
Hi guys,

I'm trying to debug & hack the Flynn project and found some wired facts, hope you guys can lead me out.

Pre: I installed Flynn by vagrant on Ubuntu 14.04 server.
>
1. After I have uploaded my app by `git push`and Flynn has built it to a slug, now how Flynn run this slug? Could anyone please provide me a typical workflow? 
2. I noticed Flynn is using `libvirt-lxc` to start container, does that mean Flynn only use image from Docker but start LXC containers instead of sending requests to docker daemon?
3. If 2 is true, how Flynn deal with some docker-specific features like `link`, `volume`?
4. I also noticed we can upload images to Flynn according on https://github.com/flynn/flynn/pull/507, but Flynn will also start LXC container to run that image instead of load it to dockerd, right?

Thanks for any tips, If there's code walk through, that'll be really helpful!!

**Comment by lmars** [comment]
@resouer 

> After I have uploaded my app by git pushand Flynn has built it to a slug, now how Flynn run this slug? Could anyone please provide me a typical workflow?

Once your application has been compiled by the buildpacks, a slug is created (which is a gzipped tarball) and uploaded to the blobstore (a simple HTTP file server), see [here](https://github.com/flynn/flynn/blob/master/slugbuilder/builder/build.sh).

Your application is then run by the slugrunner which downloads and extracts the slug, determines what command to run from the Procfile, and then runs it, see [here](https://github.com/flynn/flynn/blob/master/slugrunner/runner/init).

> I noticed Flynn is using libvirt-lxc to start container, does that mean Flynn only use image from Docker but start LXC containers instead of sending requests to docker daemon?

You are correct, Flynn currently uses images built by Docker, and runs containers using libvirt-lxc, so does not use Docker at runtime.

> If 2 is true, how Flynn deal with some docker-specific features like link, volume?

Flynn does not use the concept of links, instead every job has a cluster routable IP address and can register their address / port in service discovery. This means they can be found using either the service discovery HTTP or DNS APIs (e.g. IP addresses of PostgreSQL can be found via DNS using the hostname `postgres.discoverd`, with the primary being `leader.postgres.discoverd`).

Flynn uses ZFS to provide persistent volumes to jobs which typically get mounted at `/data` inside the container.

> I also noticed we can upload images to Flynn according on #507, but Flynn will also start LXC container to run that image instead of load it to dockerd, right?

You cannot currently upload Docker images directly to Flynn (see #1085 for my thoughts on changing that), but you can reference images from a Docker registry which will be downloaded and used to run a job (e.g. [see this Redis example](https://gist.github.com/lmars/fca0f18b4ecd507fb39a)).

Docker images are just a collection of tarballs and metadata, and you do not need Docker to use them, so yes we also use libvirt-lxc to run jobs even when pulling Docker images from a registry.

See the [pinkerton](https://github.com/flynn/flynn/tree/master/pinkerton) library for more information on how Docker images are downloaded and managed in Flynn.

**Comment by jschneiderhan** [comment]
@resouer thank you for asking, and @lmars thank you for answering in depth. I was also confused while walking through the code because I expected to see Docker API calls to create/run containers. Everything makes a lot more since now.

@lmars is there any history as to why libvirt-lxc was selected over running using the Docker daemon? I don't ask as a criticism, I'm just interested in learning the merits of each and when to use one over the other.

**Comment by lmars** [comment]
@jschneiderhan Flynn did originally use Docker as the container runtime, but at the time (late 2013 / early 2014) there were some known issues which were affecting us, and since we just needed something to run containers, we tried out libvirt-lxc and it worked well, so we switched and never looked back.

We are most likely going to revisit the container runtime and use something built-in to `flynn-host` to reduce our external dependencies further and give us maximum flexibility with their setup and configuration.

Whilst we understand users are always curious, Flynn aims to make things like these implementation details which do not affect the end-user experience.
